### PR TITLE
feat(ux): add duckdb as the default backend

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds DuckDB as the default backend to use for things that don't have a backend.

Right now, those things are:

1. `ops.Literal`s
2. `ops.InMemoryTable`s

This is useful for doing analysis on in-memory things without having to think too hard about
how to do it. The workflow is one call to `ibis.memtable` on the object and you're off to
the races.

### CODE REVIEW

**Comment by cpcloud** [comment]
I'll change the `default_backend_func` to be `ibis.config._default_backend_func`

File: ibis/backends/tests/test_client.py
Comment (kszucs) [inline ibis/backends/tests/test_client.py:675]: What happens when the user tries to execute an expression based on an unbound table?

Comment (cpcloud) [reply ibis/backends/tests/test_client.py:675]: The execution will fail.

Comment (cpcloud) [reply ibis/backends/tests/test_client.py:675]: The failure mode isn't great:

```
DBAPIError: (duckdb.Error) Invalid Error: AttributeError: 'function' object has no attribute '_export_to_c'
[SQL: SELECT t0.a
FROM t AS t0
 LIMIT ?]
[parameters: (10000,)]
(Background on this error at: https://sqlalche.me/e/14/dbapi)
```

File: ibis/expr/types/core.py
Comment (kszucs) [inline ibis/expr/types/core.py:249]: We could be a bit more verbose here and tell the user to `ibis.<backend>.execute()` or set a default backend.

Comment (kszucs) [reply ibis/expr/types/core.py:249]: Also, could you please cover it with a test case so we ensure that this error reporting remains persistent after the refactors?

Comment (cpcloud) [reply ibis/expr/types/core.py:249]: Yep!

File: ibis/backends/tests/test_client.py
Comment (cpcloud) [reply ibis/backends/tests/test_client.py:675]: This is no longer relevant due to refactoring the code to handle the case where an expression isn't executable.

File: ibis/expr/types/core.py
Comment (cpcloud) [reply ibis/expr/types/core.py:249]: Done!

**Review by kszucs — APPROVED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
does this not require an additional import?

**Comment by cpcloud** [comment]
> does this not require an additional import?

Can you clarify where you think an additional import is required?

Also, can you avoid using the "Request Changes" functionality to ask questions and reserve it only for requesting specific changes?

**Comment by jreback** [comment]
this makes duckdb a required import now

i think that is a major breaking change

**Comment by kszucs** [comment]
Could we fall back to pandas if duckdb is not available? Pandas should be a mandatory dependency, at least for now.

**Comment by jreback** [comment]
falling back to pandas seems reasonable though their is a failed import cost constantly if that happens which may not be trivial

**Comment by jreback** [comment]
not averse to in the future having duckdb as a default 

but that needs to be a separate and deliberate decision

**Comment by cpcloud** [comment]
> this makes duckdb a required import now

> i think that is a major breaking change

DuckDB is not a required import and so this isn't a breaking change.

Did you read the code? Did you notice that there are no changes to `pyproject.toml`? Can you demonstrate a scenario where this is a required import? Most of the CI jobs do **not** install duckdb, and yet somehow they continue to pass!

> but that needs to be a separate and deliberate decision

The PR **is** the deliberate decision.

If you can demonstrate a specific verifiable problem then we can continue the discussion. Otherwise, this PR is ready to merge.

**Comment by jreback** [comment]
> The PR is the deliberate decision.

where exactly is this discussion?

this is such a huge change that it needs discussion

and would be -1 at this time

**Comment by jreback** [comment]
further i think you need to prove that this isn't a major performance regression in thr lack of import 

oh and since when do we have a default backend?

**Comment by cpcloud** [comment]
> > The PR is the deliberate decision.
> 
> where exactly is this discussion?

There's no requirement to have discussion before putting up a PR, that would be ridiculously prohibitive.

If anyone has actionable objections they are welcome to describe them and comment on the PR.

> 
> this is such a huge change that it needs discussion
> 
> and would be -1 at this time

You still haven't demonstrated what the problem is. Again, **there is no requirement to have duckdb installed**. Can you describe a scenario where the import is required?

**Comment by jreback** [comment]
you are skipping the test if it's not installed

**Comment by cpcloud** [comment]
I've added tests that cover both cases, and also made it cheap to fail: the import failure overhead is incurred once per process.

After everything's green, I'm going to merge the PR.

**Comment by cpcloud** [comment]
@jreback

In general, one person's -1 is not enough to tank a PR, especially if it's already been approved. If you want to start contributing again, then your input will carry more weight.

I know that you've contributed a lot in the past, but continued interaction in some other form besides blocking PRs is what carries weight.

This applies to anyone who wants to contribute.

Adding these tests and performance enhancements is a courtesy and I will not be doing things like this going forward.

**Comment by jreback** [comment]
@cpcloud that is not a great attitude at all

you are aware we heavily rely on ibis and are in collaboration with you and others

we deeply care about the breaking changes and is it important to telegraph these and not simply override eveyone

-1 but a core maintainer should trigger discussion otherwise this is not open source

**Comment by cpcloud** [comment]
> you are aware we heavily rely on ibis and are in collaboration with you and others

I am very aware of that.

We have spent a lot of time and effort accommodating your requests to avoid going to 4.0.0 immediately after 3.1.0 to avoid a large swatch of breaking changes. We have done so by splitting our development into two separate branches, causing a lot of additional unnecessary rebase work to be done.

We did that because we value users' input and decided it was in the best interest of the project to give you more time to catch up.

I think there's been plenty of good will from our side.

All I'm asking for is a larger quantity of and higher quality maintenance work from you, in the form of actions other than blocking reviews with questions that no reasonable person can understand. We aren't going to try and guess what your concerns are.

> we deeply care about the breaking changes and is it important to telegraph these and not simply override eveyone

We also care about breaking changes otherwise we'd have released 4.0 a long time ago, but this isn't a breaking change, so there's not much to telegraph beyond a release note in the Features section of the release notes for 3.2.

> -1 but a core maintainer should trigger discussion otherwise this is not open source

Discussion is highly encouraged. Blocking PRs with a single nondescript question is not.

**Comment by cpcloud** [comment]
Alright, this has been approved and contains no breaking changes that we are aware of, with the rationale explicitly stated above, so I'm going to merge it.

We're happy to address fixes for this if they are needed in a subsequent release and we'll create patch releases for that from a branch.

After this is green, I'll cut the 3.2.0 release 🎉

**Comment by jreback** [comment]
still beating a dead horse with my objection that this is semantically changing execution 

we now can execute inbound tables implicitly - this breaks a tenant of ibis that execution must specify a backend

this is actually a slippery slope

who's to say that we should just always default to x engine and try to execute

i think this is bad user semantics and is too implicit with a very narrow good case (eg that is what u want)

the shouting is not helping anyone here and is distracting

**Pull request merged**
